ray-operator/test/e2eautoscaler/raycluster_autoscaler_test.go

```diff
@@ -194,3 +196,84 @@ func TestRayClusterAutoscalerWithCustomResource(t *testing.T) {
+	scriptsAC := newConfigMap(namespace.Name, files(test, "create_concurrent_tasks.py"))
+	scripts, err := test.Client().Core().CoreV1().ConfigMaps(namespace.Name).Apply(test.Ctx(), scriptsAC, TestApplyOptions)
+	g.Expect(err).NotTo(gomega.HaveOccurred())
+	test.T().Logf("Created ConfigMap %s/%s successfully", scripts.Namespace, scripts.Name)
+
+	setupReadyWg.Add(1)
+	taskExecWg.Add(1)
+	go func() {
+		test.T().Run("Create a RayCluster with autoscaling enabled", func(_ *testing.T) {
+			groupName := "custom-resource-group"
+
+			rayClusterSpecAC := rayv1ac.RayClusterSpec().
+				WithEnableInTreeAutoscaling(true).
+				WithRayVersion(GetRayVersion()).
+				WithHeadGroupSpec(rayv1ac.HeadGroupSpec().
+					WithRayStartParams(map[string]string{"num-cpus": "0"}).
+					WithTemplate(headPodTemplateApplyConfiguration())).
+				WithWorkerGroupSpecs(rayv1ac.WorkerGroupSpec().
+					WithReplicas(0).
+					WithMinReplicas(0).
+					WithMaxReplicas(maxReplica).
+					WithGroupName(groupName).
+					WithRayStartParams(map[string]string{"num-cpus": "1", "resources": `'{"CustomResource": 1}'`}).
+					WithTemplate(workerPodTemplateApplyConfiguration()))
+			rayClusterAC := rayv1ac.RayCluster("ray-cluster", namespace.Name).
+				WithSpec(apply(rayClusterSpecAC, mountConfigMap[rayv1ac.RayClusterSpecApplyConfiguration](scripts, "/home/ray/test_scripts")))
+
+			var err error
+			rayCluster, err = test.Client().Ray().RayV1().RayClusters(namespace.Name).Apply(test.Ctx(), rayClusterAC, TestApplyOptions)
+			g.Expect(err).NotTo(gomega.HaveOccurred())
+			test.T().Logf("Created RayCluster %s/%s successfully", rayCluster.Namespace, rayCluster.Name)
+			setupReadyWg.Done()
+
+			// Wait for RayCluster to become ready and verify the number of available worker replicas.
+			g.Eventually(RayCluster(test, rayCluster.Namespace, rayCluster.Name), TestTimeoutMedium).
+				Should(gomega.WithTransform(RayClusterState, gomega.Equal(rayv1.Ready)))
+			g.Expect(GetRayCluster(test, rayCluster.Namespace, rayCluster.Name)).To(gomega.WithTransform(RayClusterDesiredWorkerReplicas, gomega.Equal(int32(0))))
+
+			headPod, err := GetHeadPod(test, rayCluster)
+			g.Expect(err).NotTo(gomega.HaveOccurred())
+			test.T().Logf("Found head pod %s/%s", headPod.Namespace, headPod.Name)
+
+			// Create a number of tasks and wait for their completion, and a worker in the "custom-resource-group" should be created.
+			ExecPodCmd(test, headPod, common.RayHeadContainer, []string{"python", "/home/ray/test_scripts/create_concurrent_tasks.py"})
+			taskExecWg.Done()
+		})
+	}()
+
+	// Periodically check whether replica count has autoscaled to maximum value.
+	setupReadyWg.Wait()
+	ticker := time.NewTicker(200 * time.Millisecond)
+	defer ticker.Stop()
+
+	// Block until current replica has reached the preset maximum value, which proves the autoscaling works.
+	for range ticker.C {
+		pods, _ := GetWorkerPods(test, rayCluster)
+		if len(pods) == maxReplica {
+			break
+		}
+	}
```

Comment:
Use `Eventually`, `WithTransform`, and `HaveLen` instead.

Reply:
I don't think that's correct, the purpose of the test case is to check whether autoscaler could scale to the preset max replica count under great load, which *could* only happen in middle of the test, but not at the end.

Reply:
Well, `Eventually` is effectively equal to polling check.

Reply:
I think we also need to increase `autoscalerOptions.idleTimeoutSeconds` to avoid it becoming a flaky test.

Reply:
We can also set `idleTimeoutSeconds` to a large value and then check after `ExecPodCmd` completed. And then we don't need to use goroutine at all.

Reply:
Updated, though I think goroutine solution is intuitive enough :(